drivers/console/uart_console.c

```diff
@@ -335,6 +339,114 @@ static void handle_ansi(u8_t byte, char *line)
 	atomic_clear_bit(&esc_state, ESC_ANSI);
 }
 
+#ifdef CONFIG_UART_CONSOLE_MCUMGR
+
+static void clear_mcumgr(void)
+{
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_PKT);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_PKT_2);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_DATA);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_DATA_2);
+}
+
+/**
+ * These states indicate whether an mcumgr frame is being received.
+ */
+#define CONSOLE_MCUMGR_STATE_NONE       1
+#define CONSOLE_MCUMGR_STATE_HEADER     2
+#define CONSOLE_MCUMGR_STATE_PAYLOAD    3
+
+static int read_mcumgr_byte(uint8_t byte)
+{
+	bool data_1;
+	bool data_2;
+	bool pkt_1;
+	bool pkt_2;
+
+	pkt_1 = atomic_test_bit(&esc_state, ESC_MCUMGR_PKT);
+	pkt_2 = atomic_test_bit(&esc_state, ESC_MCUMGR_PKT_2);
+	data_1 = atomic_test_bit(&esc_state, ESC_MCUMGR_DATA);
+	data_2 = atomic_test_bit(&esc_state, ESC_MCUMGR_DATA_2);
+
+	if (pkt_2 || data_2) {
+		/* Already fully framed. */
+		return CONSOLE_MCUMGR_STATE_PAYLOAD;
+	}
+
+	if (pkt_1) {
+		if (byte == MCUMGR_SERIAL_HDR_PKT_2) {
+			/* Final framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_PKT_2);
+			return CONSOLE_MCUMGR_STATE_PAYLOAD;
+		}
+	} else if (data_1) {
+		if (byte == MCUMGR_SERIAL_HDR_FRAG_2) {
+			/* Final framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_DATA_2);
+			return CONSOLE_MCUMGR_STATE_PAYLOAD;
+		}
+	} else {
+		clear_mcumgr();
```

Comment:
Can you explain with a comment here under what conditions clear_mcumgr() is necessary?

It seems you have already proven that they are all unset.